### ISSUE TITLE
performance (# of recompiles): do not constant fold when it provides no benefit

### ISSUE THREAD

**Issue by tfogal** [Lightning-AI/lightning-thunder#897]
## 🚀 Feature

If there's no advantage to a constant fold, we should not perform it. Consider the following difference between two graphs:

```diff
$ diff -u graph-14.log.txt graph-30.log.txt
--- graph-14.log.txt    2024-07-31 19:06:59.118184530 +0000
+++ graph-30.log.txt    2024-07-31 19:07:05.450221326 +0000
@@ -1,4 +1,4 @@
-graph 14:
+graph 30:
 # Constructed by Delete Last Used (took 0 milliseconds)
 import torch
 from torch import Tensor
@@ -37,15 +37,15 @@
     # value_layer = ltorch.reshape(t4, [80, 384, 128])  # value_layer: "cuda:0 f16[80, 384, 128]"
       # value_layer = prims.reshape(t4, (80, 384, 128))  # value_layer: "cuda:0 f16[80, 384, 128]"
   del t4
-  matmul_result = torch.baddbmm(matmul_input_buffer, query_layer, key_layer, beta=0.0, alpha=0.08838834764831843, out=None)  # matmul_result: "cuda:0 f32[80, 384, 384]"
-    # matmul_result = ltorch.baddbmm(matmul_input_buffer, query_layer, key_layer, beta=0.0, alpha=0.08838834764831843, out=None)  # matmul_result: "cuda:0 f32[80, 384, 384]"
+  matmul_result = torch.baddbmm(matmul_input_buffer, query_layer, key_layer, beta=0.0, alpha=0.044194173824159216, out=None)  # matmul_result: "cuda:0 f32[80, 384, 384]"
+    # matmul_result = ltorch.baddbmm(matmul_input_buffer, query_layer, key_layer, beta=0.0, alpha=0.044194173824159216, out=None)  # matmul_result: "cuda:0 f32[80, 384, 384]"
       # t28 = prims.convert_element_type(query_layer, dtypes.float16)  # t28: "cuda:0 f16[80, 384, 128]"
       # t29 = prims.convert_element_type(key_layer, dtypes.float16)  # t29: "cuda:0 f16[80, 128, 384]"
       # t30 = ltorch.matmul(t28, t29)  # t30: "cuda:0 f16[80, 384, 384]"
         # t30 = prims.matmul(t28, t29)  # t30: "cuda:0 f16[80, 384, 384]"
-      # t33 = ltorch.mul(0.08838834764831843, t30)  # t33: "cuda:0 f16[80, 384, 384]"
+      # t33 = ltorch.mul(0.044194173824159216, t30)  # t33: "cuda:0 f16[80, 384, 384]"
         # t31 = prims.convert_element_type(t30, dtypes.float32)  # t31: "cuda:0 f32[80, 384, 384]"
-        # t32 = prims.mul(0.08838834764831843, t31)  # t32: "cuda:0 f32[80, 384, 384]"
+        # t32 = prims.mul(0.044194173824159216, t31)  # t32: "cuda:0 f32[80, 384, 384]"
         # t33 = prims.convert_element_type(t32, dtypes.float16)  # t33: "cuda:0 f16[80, 384, 384]"
       # t34 = ltorch.mul(0.0, matmul_input_buffer)  # t34: "cuda:0 f32[80, 384, 384]"
         # t34 = prims.mul(0.0, matmul_input_buffer)  # t34: "cuda:0 f32[80, 384, 384]"
```

[graph-14.log.txt](https://github.com/user-attachments/files/16446328/graph-14.log.txt)
[graph-30.log.txt](https://github.com/user-attachments/files/16446330/graph-30.log.txt)

Here it seems we folded in the `alpha` value, but passing the single element value instead of a constant is not going to change performance in any meaningful way, and it would mean that we could reuse the previous graph instead of generating another one.

### Motivation

[NeMo's NeVA](https://github.com/Lightning-AI/lightning-thunder/issues/343) includes 83 candidate graphs (when `dynamic=False`). These are graphs 14 and 30.

### Pitch

Ideally `alpha` would become an argument to `def computation` and just get passed in.

### Alternatives

The existing approach is just to recompile, which works but induces more recompilation than necessary.

### Additional context

cc @tfogal

**Comment by tfogal** [comment]
closing for now; this is a minor thing and actually isn't possible with `thunder.jit` alone; I'll want to create a wider design idea for dealing with duplicate graphs and circulate that, instead of one-offing each possible item that comes up.
